### PR TITLE
Effect guards

### PR DESCRIPTION
**Pull request by drbig**
Addresses #13070 - which is a _weird_ one.

This pretty much "cures the symptoms only"... which I guess is better than nothing. `[CR]` for I'm not sure we actually want such defensive mechanism(s) in.

### CODE REVIEW

File: src/effect.cpp [hunk 1]
```diff
@@ -407,6 +408,11 @@ bool effect_type::load_decay_msgs(JsonObject &jo, std::string member)
 
 std::string effect::disp_name() const
 {
+    if (eff_type->name.size() < 1) {
```
Comment (BevapDin) [inline src/effect.cpp]: Does `name` not have an `empty()` method, why not use that?

Comment (drbig) [inline src/effect.cpp]: Sure. But the bigger question is still: do we even't want this kind of guards?

Comment (BevapDin) [inline src/effect.cpp]: > Sure. But the bigger question is still: do we even't want this kind of guards?

As much as possible should be tested when the effect data is loaded. Currently `intensity` can only be 1, 2 or 3 (is that actually correct?), so the load function should check that `names` contains either nothing or exactly 3 entries. Then one only needs to check for emptiness before accessing `name[...]`. If an empty vector is not allowed at all, that should be checked (and rejected) upon load as well (and not checked here).

The code is a bit inconsistent about whether `name` can be empty. `get_speed_name` checks for an empty vector, `disp_name` does not and assumes it always non-empty. I don't know whether `name` should be allowed to be empty. If empty is not allowed, the check for it in `get_speed_name` should be removed (it's superfluous).


In short: if it's conceptually required (as in: `names` *must* have exactly three entries), it should be enforced when loading the data. Everything else should assume this requirement is already checked.

Similar with intensity. If there is an invariant (intensity must be 1, 2 or 3), it should be enforced whenever the value is changed (make sure it's changed to a valid value). Other code should simply assume it's correct. Clamping in `decay` is certainly correct.

**Comment by DanmakuDan** [comment]
The root of the problem in #13070 is that "brainworms" is set in the code as an effect, but effects.json has it as "brainworm". The mismatch leads to basically a null effect type. I think effect::decay should instead be throwing a debug message if you can reach that code with eff_type->id = "", or ignore any eff_type usage if the id is blank. I found this working backwards from the effect::decay crash in gdb, and checking what the save values were, since "brainworms" was the only blank effect in the player save.

**Comment by drbig** [comment]
Ok. So we have two things here:
1) Actual fix for #13070 which is making sure we use `brainworms` everywhere
2) Proper guards for effects, which I agree should do as much as possible on load (no idea if `intensity` has to be in `[1, 3]`, so this might take a bit of reading and/or discussion)

**Comment by BevapDin** [comment]
> The root of the problem in #13070 is that "brainworms" is set in the code as an effect, but effects.json has it as "brainworm". The mismatch leads to basically a null effect type. 

In that case. The function that loads the effects of a creature from a save (`Creature::load`?) should check that each effect id is (still) valid and ignore invalid effects (+ debug message).

The function that adds effects to creatures should also ensure that only valid effects are added.

This allows *all* the other code to assume the effect type is valid, while it creates minimal runtime overhead.

**Comment by DanmakuDan** [comment]
I checked the add_effect part with a debugger, I am guessing "" is apparently a valid effect in the effect map because it is not exiting the method like it should before it finishes adding the effect to the player.

**Comment by BevapDin** [comment]
Yes, because `Creature::add_effect` modifies the map of loaded effect types:
```C++
    // Mutate to a main (HP'd) body_part if necessary.
    if (effect_types[eff_id].get_main_parts()) {
```
Using the `operator[]` on a map creates the requested entry if it is not already contained. After this line, `eff_id` is contained in `effect_types` and is therefor per definition a valid effect id.

It should use iterators:
```C++
    const auto iter = effect_types.find( eff_id );
    if( iter == effect_types.end() ) {
        debugmsg( ... );
    }
    const effect_type *type = &iter->second;
```

Also: `effect` should store a pointer to a const `effect_type` object.

**Comment by drbig** [comment]
I wonder which of these 4 checks are really needed :/

**Comment by drbig** [comment]
Or we can take the approach of adding all checks, and removing some if someone proves they are (at that point) actually not needed.

**Comment by DanmakuDan** [comment]
I think all of them can be relevant, since people can import older saves that were crashing before. The intensity clamping probably needs a debug message since it shouldn't be reaching effect::decay without a valid value.

**Comment by drbig** [comment]
@DanmakuDan I think I tried that but it made getting past the debug spam really tedious.
